### PR TITLE
feat: GitHub provider

### LINKED ISSUE #13: Read Notes Repo From GitHub
It would be great if we could get a notes repository on request from GitHub and use the latest state of the repo to construct the graph.

A/C can look something like:

1. As a user of the mindgraph library, I am able to provide a URL to my obsidian notes repo on GitHub (or GitLab? or Bitbucket?) and an API token in a secure server-side environment.
2. Mindgraph can fetch that information on demand and create GraphData - similar to how `readFromFileSystem` does today.

### PR DESCRIPTION
**Pull request by jollyjerr**
This PR closes #13 

## What does this PR do?

Adds the ability to read your notes from GitHub directly. I didn't add any caching logic or anything for now until we decide on the direction for mindgraph to move as a whole. As a consumer of this package, you could cache the result.

I added tests that assert the GitHub and FileSystem providers have the same, correct, output.

There is more common code that could be refactored out of the providers and also some requests could be parallelized, but I didn't really feel like performance tuning or totally abstracting stuff at this stage.

### CODE REVIEW

File: packages/read/src/common.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+export function extractLinksFromLine(
+  line: string,
+  filePath: string,
+): GraphData['links'] {
+  const formattedLinks = [];
+  const links = line.match(LINK_CONTENT_REGEX) || [];
+
+  for (const link of links) {
+    const path = LINK_CONTENT_REGEX.exec(link)?.at(1);
```
Comment (mpryor) [inline packages/read/src/common.ts:29]: I think there are some unintended consequences of using this `exec` along with the global flag on this particular Regex, see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description

I wrote a fix (I think), here: https://github.com/jollyjerr/mindgraph/pull/29/files#diff-09c58197e1dfd80696e23cc9172cbf7fe3cc9f9c048c349a133f49256705e4edR26, and also added some unit tests for this function

File: packages/read/src/common.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (is_valid_link_path(path)) {
+      const linkDirections = path.split('/');
+      const pathToTargetFile = filePath.split('/');
+
+      pathToTargetFile.pop();
+
+      for (const direction of linkDirections) {
```
Comment (mpryor) [inline packages/read/src/common.ts:37]: nit: I thought the name of `direction` and `linkDirections` was kinda confusing. I'm wondering if `part` and `linkPathParts` or something might be better?

File: packages/read/src/common.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        target: pathToTargetFile.join('/'),
+      });
+    }
+  }
+
+  return formattedLinks;
+}
+
+function is_valid_link_path(path: string | undefined): path is string {
+  return !!(path && !path.includes('://') && path.includes(MARKDOWN_EXTENSION));
```
Comment (mpryor) [inline packages/read/src/common.ts:60]: nit: `MARKDOWN_EXTENSION` has a constant, but `://` is left as a magic string. Maybe a constant like PROTOCOL_DELIMITER

File: packages/read/src/file-system/file-system.test.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+describe('FileSystem provider', () => {
+  it('should produce the expected graph', async () => {
+    const graph = await FileSystem.read({
+      path: path.resolve(__dirname, '../../../', 'draw', 'example', 'notes'),
+    });
+
+    expect(formatGraphForTestSnapshot(graph)).toMatchFileSnapshot(
```
Comment (mpryor) [inline packages/read/src/file-system/file-system.test.ts:15]: No suggestions, just thought the testing approach of building a graph from the filesystem, and using the same graph against the Github API was pretty clever! Good work!

File: packages/read/src/github/github.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      owner,
+      repo,
+      ref: `heads/${branch}`,
+    });
+
+    return build_graph(octo, { owner, repo, path: basePath, ref });
+  },
+};
+
+type GitHubResult = {
```
Comment (mpryor) [inline packages/read/src/github/github.ts:40]: Question: I'm curious about the internal definition of a `GithubResult` versus something leveraging Octokit's openapi types? https://github.com/octokit/types.ts/issues/267#issuecomment-781637455

Not looking for anything to be changed, just curious.

**Review by mpryor — APPROVED** [review]
This is dope! Just some nits

File: packages/read/src/common.ts (hunk 1 above)
Comment (jollyjerr) [reply packages/read/src/common.ts:29]: Woah I had no idea this statefulness was a thing 😮, I'll merge your new implementation for sure! Thank you for doing that!

File: packages/read/src/github/github.ts (hunk 5 above)
Comment (jollyjerr) [reply packages/read/src/github/github.ts:40]: Ya idk lol it's already typed from octokit itself but the type was annoying and I couldn't get it to work so I just wrote a quick one that had what I cared about. Their openapi types would also do the trick

Screenshot 2023-07-03 at 6 28 36 PM

**Pull request merged**